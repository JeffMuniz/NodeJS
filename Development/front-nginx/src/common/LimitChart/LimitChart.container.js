import React, { Component } from "react";
import { connect } from "react-redux";
import { number, shape, func, string, bool } from "prop-types";
import * as companyTreeActions from "src/redux/modules/selectedCompanyTree/actions/selectedCompanyTree";
import View from "./LimitChart";

export class LimitChartContainer extends Component {
  state = {
    stateIdGroup: 0,
  };

  componentDidMount() {
    this.getLimit();
  }

  componentDidUpdate() {
    const { stateIdGroup } = this.state;
    const {
      selectedGroup: { id: idGroup },
    } = this.props;

    if (idGroup && idGroup !== stateIdGroup) this.getLimit();
  }

  getLimit = async () => {
    const {
      getOrdersLimit,
      userId,
      selectedGroup: { id: idGroup },
    } = this.props;

    this.setState({ stateIdGroup: idGroup }, async () => {
      if (idGroup) {
        await getOrdersLimit({ idGroup, userId });
      }
    });
  };

  render() {
    const {
      hasBorderTop,
      selectedGroup: {
        ordersLimit: {
          percentage,
          availableValue,
          usedLimit,
          totalLimit,
          checkForLimit,
        },
      },
    } = this.props;

    if (!checkForLimit) return null;

    return (
      <View
        hasBorderTop={hasBorderTop}
        percentage={percentage}
        availableValue={availableValue}
        usedLimit={usedLimit}
        totalLimit={totalLimit}
      />
    );
  }
}

export const mapStateToProps = ({
  selectedCompanyTree,
  user: {
    profile: { data: userData },
  },
}) => ({
  userId: userData && userData.id,
  selectedGroup: selectedCompanyTree.selectedGroup,
});

export const mapDispatchToProps = {
  getOrdersLimit: companyTreeActions.getSelectedGroupOrdersLimit,
};

LimitChartContainer.propTypes = {
  selectedGroup: shape({
    ordersLimit: shape({
      percentage: number.isRequired,
      availableValue: number.isRequired,
      usedLimit: number.isRequired,
      totalLimit: number.isRequired,
    }),
  }).isRequired,
  getOrdersLimit: func.isRequired,
  userId: string.isRequired,
  hasBorderTop: bool.isRequired,
};

export default connect(
  mapStateToProps,
  mapDispatchToProps,
)(LimitChartContainer);
