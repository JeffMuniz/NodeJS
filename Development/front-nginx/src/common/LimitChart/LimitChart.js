import React, { Component } from "react";
import { string, oneOfType, number, bool } from "prop-types";
import {
  dustyGray,
  redFlagGetnet,
  darkGreen,
  chartOrange,
  chartGrey,
} from "@colors";
import { If, toMoneyMask } from "@utils";

import {
  Container,
  ChartWrapper,
  ChartTransform,
  Chart,
  Percentage,
  BottomLabel,
  DescriptionWrapper,
  Title,
  AvailableValue,
  Description,
  UsedLimit,
  TotalLimit,
  MailTo,
} from "./LimitChart.styles";

class LimitChart extends Component {
  constructor(props) {
    super(props);

    this.state = {
      color: this.getColor(props.percentage),
      background: this.getBackground(props.percentage),
      statePercentage: props.percentage,
    };
  }

  componentDidUpdate() {
    const { percentage } = this.props;
    const { statePercentage } = this.state;
    if (percentage === statePercentage) return;
    this.setStateColor(percentage);
  }

  setStateColor = percentage => {
    this.setState({
      color: this.getColor(percentage),
      background: this.getBackground(percentage),
      statePercentage: percentage,
    });
  };

  getColor = percentage => {
    const colors = {
      "0": darkGreen,
      "25": dustyGray,
      "50": chartOrange,
      "75": redFlagGetnet,
      "100": redFlagGetnet,
    };

    const color = Object.keys(colors)
      .filter(x => Number(x) <= Number(percentage))
      .pop();

    return colors[color];
  };

  getBackground = percentage =>
    Number(percentage) >= 50 ? chartGrey : darkGreen;

  render() {
    const { color, background } = this.state;
    const {
      percentage,
      availableValue,
      usedLimit,
      totalLimit,
      hasBorderTop,
    } = this.props;

    const contactEmail = "contasareceber@macvisacard.com.br";
    const totalLimitUsed = percentage >= 100;

    return (
      <Container
        id="lmc_container"
        hasBorderTop={hasBorderTop}
        aria-percentage={percentage}
        aria-used-color={color}
        aria-left-color={background}
      >
        <ChartWrapper>
          <ChartTransform>
            <Chart
              data={[{ value: 1, key: 1, color }]}
              reveal={percentage > 100 ? 100 : percentage}
              lineWidth={25}
              percentage={percentage}
              lengthAngle={270}
              rounded
              background={background}
            />
          </ChartTransform>
          <Percentage id="lmc_percent">{percentage}%</Percentage>
          <BottomLabel>Utilizado</BottomLabel>
        </ChartWrapper>
        <DescriptionWrapper>
          <Title id="lmc_your_limit">
            seu limite disponível para pedidos é:
          </Title>
          <If test={availableValue <= 0}>
            <AvailableValue id="lmc_available_value">
              {toMoneyMask(0)}
            </AvailableValue>
          </If>
          <If test={availableValue > 0}>
            <AvailableValue id="lmc_available_value">
              {toMoneyMask(availableValue)}
            </AvailableValue>
          </If>
          <If test={!totalLimitUsed}>
            <Description id="lmc_desc">
              Você está utilizando
              <UsedLimit id="lmc_used_value">
                {` ${toMoneyMask(usedLimit)} `}
              </UsedLimit>
              do seu limite total de
              <TotalLimit id="lmc_total_value">
                {` ${toMoneyMask(totalLimit)} `}
              </TotalLimit>
            </Description>
          </If>
          <If test={totalLimitUsed}>
            <Description id="lmc_desc_mailto">
              entre em contato com
              <MailTo href={`mailto:${contactEmail}`}>
                &nbsp;{contactEmail}
              </MailTo>
            </Description>
          </If>
        </DescriptionWrapper>
      </Container>
    );
  }
}

LimitChart.propTypes = {
  percentage: oneOfType([string, number]).isRequired,
  availableValue: number.isRequired,
  usedLimit: number.isRequired,
  totalLimit: number.isRequired,
  hasBorderTop: bool.isRequired,
};

export default LimitChart;
