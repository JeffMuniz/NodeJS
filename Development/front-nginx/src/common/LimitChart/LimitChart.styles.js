import styled from "styled-components";
import { Row } from "react-styled-flexboxgrid";
import PieChart from "src/modules/PieChart/PieChart";
import { ifStyle } from "@utils";
import {
  black,
  red,
  lighterBlack,
  simpleDarkenGrey,
  darkGreen,
  blue,
  rgbaStrongShadowGrey,
} from "@colors";

const hasBorderTop = ifStyle("hasBorderTop");

export const Container = styled(Row)`
  border-top: ${hasBorderTop(`1px solid ${rgbaStrongShadowGrey}`, "none")};
  display: flex;
  padding: 40px 0 24px;
  margin: 0 0 32px;
`;

export const ChartWrapper = styled.div`
  width: 80px;
  height: 80px;
  position: relative;
`;

export const ChartTransform = styled.div`
  transform: scaleX(-1);
`;

export const Chart = styled(PieChart)`
  transform: rotate(135deg);
`;

export const Percentage = styled.span`
  position: absolute;
  left: 17px;
  top: 30px;
  width: 45px;
  font-weight: 700;
  text-align: center;
  letter-spacing: 0.1;
`;

export const BottomLabel = styled.span`
  position: absolute;
  font-size: 12px;
  line-height: 16px;
  text-align: center;
  letter-spacing: 0.1px;
  color: ${black};
  top: 73px;
  left: 15px;
`;

export const DescriptionWrapper = styled.div`
  display: flex;
  flex-direction: column;
  margin: auto 18px;
`;
export const Title = styled.span`
  font-size: 12px;
  line-height: 26px;
  letter-spacing: 0.1px;
  color: ${red};
`;
export const AvailableValue = styled.span`
  font-weight: 600;
  font-size: 32px;
  line-height: 40px;
  letter-spacing: 0.3px;
  color: ${lighterBlack};
`;
export const Description = styled.span`
  font-weight: bold;
  font-size: 12px;
  line-height: 18px;
  letter-spacing: 0.1px;
  color: ${simpleDarkenGrey};
`;

export const UsedLimit = styled(Description)`
  color: ${darkGreen};
`;

export const TotalLimit = styled(Description)`
  color: ${black};
`;

export const MailTo = styled.a`
  color: ${blue};
  font-weight: normal;
  text-decoration: none;
`;
