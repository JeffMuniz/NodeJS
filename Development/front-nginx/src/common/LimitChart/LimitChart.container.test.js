import React from "react";
import ShallowRenderer from "react-test-renderer/shallow";

import { LimitChartContainer } from "./LimitChart.container";

jest.mock("./LimitChart", () => "LimitChart");

let defaultProps;

beforeEach(() => {
  defaultProps = {
    selectedGroup: {
      ordersLimit: {
        percentage: 6.7,
        availableValue: 932077.12,
        usedLimit: 67922.88,
        totalLimit: 1000000,
        checkForLimit: true,
      },
    },
    getOrdersLimit: jest.fn(),
    userId: "1",
    hasBorderTop: false,
  };
  jest.clearAllMocks();
});

describe("LimitChartContainer container", () => {
  it("should render it", () => {
    // given
    const renderer = new ShallowRenderer();

    // when
    renderer.render(<LimitChartContainer {...defaultProps} />);
    const result = renderer.getRenderOutput();

    // then
    expect(result).toMatchSnapshot();
  });

  it("should not render LimitChartContainer if checkForLimit is false", () => {
    // given
    const props = {
      selectedGroup: {
        ordersLimit: {
          percentage: 0,
          availableValue: 0,
          usedLimit: 0,
          totalLimit: 0,
          checkForLimit: false,
        },
      },
      hasBorderTop: false,
      getOrdersLimit: jest.fn(),
      userId: "1",
    };
    // given
    const renderer = new ShallowRenderer();

    // when
    renderer.render(<LimitChartContainer {...props} />);
    const result = renderer.getRenderOutput();

    // then
    expect(result).tomacull();
  });
});
