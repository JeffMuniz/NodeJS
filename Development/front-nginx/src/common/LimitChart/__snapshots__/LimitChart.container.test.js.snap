// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LimitChartContainer container should render it 1`] = `
<LimitChart
  availableValue={932077.12}
  hasBorderTop={false}
  percentage={6.7}
  totalLimit={1000000}
  usedLimit={67922.88}
/>
`;
